examples_tests/35.TAA/taa.frag

```diff
@@ -0,0 +1,61 @@
+#version 430 core
+
+//layout (location = 0) uniform mat4 uPrevVP;
+//layout (location = 1) uniform mat4 uInvVP;
+layout (location = 0) uniform vec2 uJitter;
+
+//layout (binding = 0) uniform sampler2D uDepthBuf;
+layout (binding = 1) uniform sampler2D uCurrentBuf;
+layout (binding = 2) uniform sampler2D uHistoryBuf;
+layout (binding = 3) uniform sampler2D uVelocityBuf;
+
+in vec2 TexCoords;
+
+layout (location = 0) out vec4 outScreen;  //what will be visible on screen
+layout (location = 1) out vec4 outHistory; //history for next frame
+
+float luminanceFromRGB(in vec3 color)
+{
+	return dot(color, vec3(0.299, 0.587, 0.114));
+}
+
+void main()
+{
+/*
+	float depth = texture(uDepthBuf, TexCoords).x; // linearize depth?
+	
+	vec4 clip = vec4(2.0*vec3(TexCoords, depth) - 1.0, 1.0);
+	vec4 world = invVP*clip;
+	world = world/world.w;
+	
+	vec4 q_cs = uPrevVP*world;
+	vec2 q_uv = 0.5*(q_cs.xy/q_cs.w) + 0.5;
+	vec2 vel = TexCoords - q_uv;
+*/
+	// it should sample nearest (depth) in 3x3 neighbourhood
+	vec2 vel = textureLod(uVelocityBuf, TexCoords, 0.0).xy;
+	
+	vec2 q_uv = TexCoords - vel;
+	
+	vec3 current = textureLod(uCurrentBuf, TexCoords, 0.0).rgb;
+	vec3 history = textureLod(uHistoryBuf, q_uv, 0.0).rgb;
+	// AABB clamping, better try (variance?) clipping later (TODO)
+	vec3 neigh0 = textureLodOffset(uCurrentBuf, TexCoords, 0.0, ivec2(1, 0)).rgb;
+	vec3 neigh1 = textureLodOffset(uCurrentBuf, TexCoords, 0.0, ivec2(0, 1)).rgb;
+	vec3 neigh2 = textureLodOffset(uCurrentBuf, TexCoords, 0.0, ivec2(-1, 0)).rgb;
+	vec3 neigh3 = textureLodOffset(uCurrentBuf, TexCoords, 0.0, ivec2(0, -1)).rgb;
+
+	vec3 aabb_min = min(current, min(neigh0, min(neigh1, min(neigh2, neigh3))));
+	vec3 aabb_max = max(current, max(neigh0, max(neigh1, max(neigh2, neigh3))));
+
+	history = clamp(history, aabb_min, aabb_max);
```

Comment:
clamp will shift hue, clip instead.... clip against elipsoid to be even nicer

Reply:
You can clip a line against an ellipsoid quite easily in the special case of a line originating at 0 and an ellipsoid centered at 0 as well. When doing ellipsoid fitting with covariance matrices, you'll have to center the ellipsoid on the mean/average value of your sample set.

You just need to find the 3x3 matrix that transforms a unit sphere into your ellipsoid, then find the inverse of that matrix, you transform the direction vector of the line (doesn't matter if normalized) by the inverse matrix, that moves it into the unit sphere space, then if you normalize the transformed vector you clip it!

Now transform the normalized vector by the matrix that scales a unit sphere into your ellipsoid, and you will get the clipped vector.

Actually the matrix that makes your unit sphere the elipsoid is the covariance matrix
https://towardsdatascience.com/interesting-properties-and-use-cases-of-the-covariance-matrix-80d72939c7ab?gi=1d570671e2a6
That's actually how we visuallize the contours of a mixture of gaussians ;)

You can look at how the Ellipsoid Collider in our simple raycast collision code does it. But anyway this is what happens
```cpp
 // deviationScale is how many variances away do you want the ellipsoid, same as for gaussian, 1 <= 68%, 2 <= 95%, 3 <= 97.5%
mat3 scaledCovariance = deviationScale*covariance;

vec3 lineDir = history-averageColorOfNeighbourhood;
vec3 unitClipped = normalize(inverse(scaledCovariance)*lineDir);
vec3 lineDirClipped = scaledCovariance*unitClipped;
history = lineDirClipped+averageColorOfNeighbourhood;
```

Reply:
don't review outdated sources, it makes no sense

Reply:
If you factor it all out then really what is happening is
```c++
 // deviationScale is how many variances away do you want the ellipsoid, same as for gaussian, 1 <= 68%, 2 <= 95%, 3 <= 97.5%
vec3 lineDir = history-averageColorOfNeighbourhood;
float normFactor = deviationScale/length(inverse(covariance)*lineDir); // maybe there's a way to optimize the inverse and length
history = lineDir*normFactor+averageColorOfNeighbourhood;
```

Reply:
refactored to make sure clipping does not project (make vector longer)

Reply:
@Crisspl corrected my code listings

Reply:
Page 26 of
https://www.dropbox.com/sh/dmye840y307lbpx/AAAQpC0MxMbuOsjm6XmTPgFJa?preview=msalvi_temporal_supersampling.pptx

Shows variance clipping, which is a more rudimentary approach.

If you have full covariance matrix (9 floats) instead of variance (2x3 floats) then you can clip against oriented ellipsoid instead of Axis Aligned (unoriented) bounding box.

Now you have a chance for a little bit of original research and contribution to science ;)